service/common/src/main/java/org/apache/polaris/service/events/listeners/FileBufferPolarisPersistenceEventListener.java

```diff
@@ -0,0 +1,276 @@
+        bufferShard.output.flush();
+    }
+
+    private void createBuffersForRealm(String realmId) {
+        HashMap<Integer, BufferShard> bufferShardsForRealm = new HashMap<>();
+        buffers.put(realmId, bufferShardsForRealm);
+        for (int i = 0; i < shardCount; i++) {
+            bufferShardsForRealm.put(i, createShard(realmId, i));
+        }
+        ScheduledFuture<?> future = threadPool.schedule(new BufferListingTask(shardCount, realmId), timeToFlush, TimeUnit.MILLISECONDS);
+    }
+
+    private BufferShard createShard(String realmId, int shardNum) {
+        File file = new File(getBufferDirectory(realmId, shardNum) + "buffer_timestamp-" + System.currentTimeMillis());
+        File parent = file.getParentFile();
+        if (parent != null && !parent.exists()) {
+            parent.mkdirs();  // Creates all missing parent directories
+        }
+        try {
+            file.createNewFile();
+            Output output = new Output(new FileOutputStream(file));
+            return new BufferShard(null, file, output);
+        } catch (IOException e) {
+            LOGGER.error("Buffer shard {} was unable to initialize: {}", shardNum, e);
+        }
+        return null;
+    }
+
+    private String getBufferDirectory(String realmId, int shardNum) {
+        return System.getProperty("java.io.tmpdir") + "/" + realmId + "/polaris-event-buffer-shard-" + shardNum + "/";
+    }
+
+    private RealmContext getRealmContext(String realmId) {
+        return () -> realmId;
+    }
+
+    private void discoverAndPushOrphanedFiles() {
+        LOGGER.info("Attempting to discover orphaned event buffer files");
+        File topLevelDirectory = new File(System.getProperty("java.io.tmpdir"));
+        File[] realmFiles = topLevelDirectory.listFiles();
+        if (realmFiles == null || realmFiles.length == 0) {
+            LOGGER.info("No discovered orphaned event buffer files in {}", topLevelDirectory.getAbsolutePath());
+            return;
+        }
+        for (File realmFile : realmFiles) {
+            LOGGER.info("Discovered orphaned event buffer potential realm: {}", realmFile.getName());
+            File[] realmBufferFiles = realmFile.listFiles();
+            if (realmBufferFiles != null) {
+                for (File shardDir : realmBufferFiles) {
+                    if (!shardDir.getName().startsWith("polaris-event-buffer-shard-")) {
+                        continue;
+                    }
+                    for (File bufferFile : shardDir.listFiles()) {
+                        LOGGER.info("Queued file flush for orphaned event buffer file: {}", bufferFile.getAbsolutePath());
+                        FileFlushTask task = new FileFlushTask(bufferFile.getAbsolutePath(), realmFile.getName());
+                        Future<?> future = threadPool.submit(task);
+                        if (createdFlushFutures.putIfAbsent(bufferFile.getAbsolutePath(), future) != null) {
```

Comment:
You shouldn't submit the future if the absolute path is already in `createdFlushFutures`. You can simply do:
```
createdFlushFutures.putIfAbsent(bufferFile.getAbsolutePath(), threadPool.submit(task))
```

Reply:
If I'm not mistaken, with your suggestion, `threadPool.submit(task)` will be run even if `putIfAbsent` returned a non-null value. Don't we need to cancel the future we've created in that case? Or else, we will have double tasks for the same file.